### PR TITLE
[table info][1/4] separate indexer async v2 db from aptosdb

### PR DESCRIPTION
**Pull request by jillxuu**
### This PR
1. remove indexer async v2 db from aptosdb
2. remove indexer async v2 db read and write from dbreader & dbwriter
3. expose indexer async v2 db reader trait
4. add indexer async v2 db reader trait to fn api and indexer fn

### Goal
https://www.notion.so/aptoslabs/Internal-Indexer-Deprecation-Next-Steps-824d5af7f16a4ff3aeccc4a4edee2763?pvs=4
1. Migrating internal indexer off of aptos db critical path, move it into its own standalone runtime service.
5. Still provide the table info mapping to both API and indexer services.
6. improve table info parsing performance to unblock indexer perf bottleneck

### Context
This effort is broken down into two parts:
1. part 1: https://github.com/aptos-labs/aptos-core/pull/10783. Add moving the table info service out of the critical path and convert it into multithread to concurrently process the request.
2. part 2 is this PR. Now we have the table info service, but this service will only be enabled by a handful FNs, so when a new FN wants to join the network but they don't want to sync from the genesis, they should be able to restore the db from a cloud service. In order to provide such a cloud service for others to download the db snapshot, this pr focuses on two things: backup and restore. backup is optional in the config, but restore logic is written in the code.

| before | after |
| --- | --- |
| Screenshot 2023-11-15 at 10 15 01 PM | Screenshot 2024-01-25 at 11 14 05 AM |

### Detailed Changes
Screenshot 2024-01-25 at 11 56 53 AM

#### backup based on epoch or transaction version? what frequency?
Pros of backup using version:
1. We have more control over the backup frequency as frequency is tunable. 
Cons of backup using version:
1. Overhead of managing and comparing backed up versions and current processing versions

Pros of backup using epoch:
1. When to backup logic is much cleaner and less error prone
Cons of backup using epoch:
1. not very configurable but still tunable by setting frequency on how many epochs behind

Decided to use epoch, because on testnet we have little over 10k epoch, divided by total txns, it gives us on average 70k txns per epoch, this sounds about the frequency of txns we'd like to backup.

#### when to restore
Decided to restore when both conditions are met: 
1. the difference btw next versions to be processed and current version from ledger is greater than a `version_diff`.
2. and time difference btw last restored timestamp in db and current timestamp is greater than a `RESTORE_TIME_DIFF_SECS`.
This is to prevent fullnode crashlooping and constantly try to restore without luck, and when version difference is not that big we don't need to spam the gcs service but rather directly state syncing from that close to head version.

#### structure of the gcs bucket
I followed the similar structure as of indexer's filestore, where we keep a `metadata.json` file in the bucket to keep track of the chain id and newest backed up epoch. and then a files folder to keep all the epoch based backup. Each db snapshot is first compressed into a tar file from folder, and then gzipped to compress to the best size possible. based on larry's point, alternative compression like bzip2 is less performant.

#### threads
Using a separate thread for backup only, base on the past experience, gcs upload could be as slow as minutes.

#### gcs prunning
Couple options we could pursue, since each backup file is a full db backup, 
1. create another service to constantly clean up the backup files
2. use gcs own policy to delete files based on time, and other conditions
3. programmatically delete old files while upload
7. constantly writing to the same file

Decided to go with gcs own policy with the proper configuration setup in the gcs deployment. Reason behind is that deploying and maintaining another service is overhead and costs more money, especially this service's responsibility is very singular; writing code for gcs object deletion is not ideal, since we're writing and deleting, need to handle different multitude of edge cases; constantly writing to the same file is def not gonna work, since gcs has strict limitation on single object write limit, only once per second.

### Test Plan
1. passing all the written unit tests on file system operation
8. locally tested and verified backup and restore, as well as table info read

### Concerns
There's a bottleneck on the size of db snapshot. Currently this db on testnet is around 250mb, based on the nature of db, the compression could get it to be 50-150mb per db snapshot. It's still too big to upload to gcs as its too slow. 

#### E2E test
from rustie:
1. set up a long-running fullnode in a new k8s project. We can use the same data-staging-us-central1 cluster, but use a new namespace, like indexer-fullnode-testnet-test  or something. This is to isolate it from everything else, but we can use the same cluster for simplicity
9. Set up a job in the same namespace that does the backup to GCS.
10. We can set up a continuous job in aptos-core CI that:
4.1. Spins up a fullnode based on the latest build. This would be the latest main nightly for instance
3.2. We can quit immediately after verifying that the restore was successful
3.3. The cost should be manageable, assuming that the restore process is quick.

#### load test
Couple things i want to verify with load testing
1. 10 fns boostrapping, can restore work for all of them
2. fns keep crashlooping, is gcs spammed based on egress & ingress
3. when file gets bigger, will backup still work and how long it could take

### CODE REVIEW

File: api/types/src/convert.rs [hunk 1]
```diff
@@ -451,19 +471,33 @@ impl<'a, R: ModuleResolver + ?Sized> MoveConverter<'a, R> {
...
+                // If table_info_reader exists, use it to get table_info
+                Ok(table_info_reader.get_table_info(handle)?)
+            } else if self.db.indexer_enabled() {
+                // If indexer is enabled in db, get table_info from the db
+                Ok(Some(self.db.get_table_info(handle)?))
+            } else {
+                // If neither condition is met, return None
+                Ok(None)
+            }
+        };
```
[the remark this answers is not in the document]
Comment (bowenyang007) [reply api/types/src/convert.rs:486]: @jillxuu seems like this needs to be drier hmm...

File: storage/schemadb/tests/db.rs [hunk 2]
```diff
@@ -20,7 +20,7 @@ define_schema!(TestSchema1, TestField, TestField, "TestCF1");
 define_schema!(TestSchema2, TestField, TestField, "TestCF2");
 
 #[derive(Debug, Eq, PartialEq)]
-struct TestField(u32);
+pub struct TestField(u32);
```
[the remark this answers is not in the document]
Comment (jillxuu) [reply storage/schemadb/tests/db.rs:23]: they need to be accessed outside of its own crate, cuz we expose db initialization in a different service, instead of the db storage crate. but they will be eventually reverted back after we deleting the old indexer db.

File: storage/aptosdb/src/db/include/aptosdb_reader.rs [hunk 3]
```diff
@@ -1026,26 +999,8 @@ impl AptosDB {
...
-    ) -> Result<Option<TableInfo>> {
-        match &self.indexer_async_v2 {
-            Some(indexer_async_v2) => indexer_async_v2.get_table_info_with_retry(handle),
-            None => bail!("Indexer Async V2 not enabled."),
-        }
-    }
-
     /// TODO(jill): deprecate Indexer once Indexer Async V2 is ready
-    fn get_table_info_from_indexer(&self, handle: TableHandle) -> Result<Option<TableInfo>> {
+    fn get_table_info_option(&self, handle: TableHandle) -> Result<Option<TableInfo>> {
```
[the remark this answers is not in the document]
Comment (jillxuu) [reply storage/aptosdb/src/db/include/aptosdb_reader.rs:1003]: we don't keep reading from old db indexer if they opt in for the new db indexer, but still want to keep the old one available for a month, meanwhile, we run [e2e test](https://www.notion.so/aptoslabs/Indexer-Replace-fullnode-API-with-indexer-APIs-bbd369a641d44920aecbf835a4939bd5?pvs=4#ef3be7a925394a7588d8f52b09e06669) daily to verify backup restore logic and dashboards for monitoring.

File: api/types/src/convert.rs [hunk 4]
```diff
@@ -422,19 +429,32 @@ impl<'a, R: ModuleResolver + ?Sized> MoveConverter<'a, R> {
...
```
[the remark this answers is not in the document]
Comment (jillxuu) [reply api/types/src/convert.rs:444]: cuz now we have two db readers since new indexer db is no longer part of the aptosdb. no matter what, we need to check two sources, but i could def improve on readability

File: storage/aptosdb/src/db/include/aptosdb_reader.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (msmouse) [reply storage/aptosdb/src/db/include/aptosdb_reader.rs:1003]: > but still want to keep the old one available for a month, meanwhile

I mean, if you stopped writing to it, the info becomes stale? Or did I read it wrong?

File: storage/schemadb/tests/db.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (msmouse) [reply storage/schemadb/tests/db.rs:23]: I mean, where? Why do you need TestField outside of this crate?

[the remark this answers is not in the document]
Comment (jillxuu) [reply storage/schemadb/tests/db.rs:23]: we don't use `TestField` directly, but i think `TestField` requires `define_schema` and we use `define_schema` in a different cargo.

File: storage/aptosdb/src/db/include/aptosdb_reader.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jillxuu) [reply storage/aptosdb/src/db/include/aptosdb_reader.rs:1003]: we have two fullnodes on indexer, we could technically use one node to write to old, one node to write to new

**Review by msmouse — APPROVED** [review]

File: storage/aptosdb/src/db/include/aptosdb_reader.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (msmouse) [reply storage/aptosdb/src/db/include/aptosdb_reader.rs:1003]: nvm. I thought the interfaces you removed from DbWriter prevents writing to the old internal indexer.

**Review by bowenyang007 — APPROVED** [review]
